presto-hive/src/main/java/com/facebook/presto/twitter/hive/thrift/ThriftGenericRow.java

```diff
@@ -0,0 +1,305 @@
+            case TType.DOUBLE:
+                return iprot.readDouble();
+            case TType.STRING:
+                return iprot.readString();
+            case TType.STRUCT:
+                return readStruct(iprot);
+            case TType.LIST:
+                return readList(iprot);
+            case TType.SET:
+                return readSet(iprot);
+            case TType.MAP:
+                return readMap(iprot);
+            default:
+                TProtocolUtil.skip(iprot, type);
+                return null;
+        }
+    }
+
+    private Object readStruct(TProtocol iprot) throws TException
+    {
+        ThriftGenericRow elem = new ThriftGenericRow();
+        elem.read(iprot);
+        elem.parse();
+        return elem;
+    }
+
+    private Object readList(TProtocol iprot) throws TException
+    {
+        TList ilist = iprot.readListBegin();
+        List<Object> listValue = new ArrayList<>();
+        for (int i = 0; i < ilist.size; ++i) {
+            listValue.add(readElem(iprot, ilist.elemType));
+        }
+        iprot.readListEnd();
+        return listValue;
+    }
+
+    private Object readSet(TProtocol iprot) throws TException
+    {
+        TSet iset = iprot.readSetBegin();
+        List<Object> setValue = new ArrayList<>();
+        for (int i = 0; i < iset.size; ++i) {
+            setValue.add(readElem(iprot, iset.elemType));
+        }
+        iprot.readSetEnd();
+        return setValue;
+    }
+
+    private Object readMap(TProtocol iprot) throws TException
+    {
+        TMap imap = iprot.readMapBegin();
+        Map<Object, Object> mapValue = new HashMap<>();
+        for (int i = 0; i < imap.size; ++i) {
+            mapValue.put(readElem(iprot, imap.keyType), readElem(iprot, imap.valueType));
+        }
+        iprot.readMapEnd();
+        return mapValue;
+    }
+
+    public Object getFieldValueForThriftId(short thriftId)
+    {
+        return values.get(thriftId);
+    }
+
+    public ThriftGenericRow deepCopy()
+    {
+        return new ThriftGenericRow(values);
+    }
+
+    public void clear() {}
+
+    public Fields fieldForId(int fieldId)
+    {
+        return new Fields((short) fieldId, "dummy");
+    }
+
+    public Object getFieldValue(Fields field)
+    {
+        return values.get(field.thriftId);
+    }
+
+    public boolean isSet(Fields field)
+    {
+        return values.containsKey(field.getThriftFieldId());
+    }
+
+    public void setFieldValue(Fields field, Object value)
+    {
+        values.put(field.getThriftFieldId(), value);
+    }
+
+    public void write(TProtocol oprot) throws TException
+    {
+        oprot.writeStructBegin(new TStruct("dummy"));
+        values.entrySet().stream()
+                         .filter(v -> v.getValue() != null)
+                         .forEach(v -> writeField(v, oprot));
+        oprot.writeFieldStop();
+        oprot.writeStructEnd();
+    }
+
+    private void writeField(Map.Entry<Short, Object> field, TProtocol oprot)
```

Comment:
Could we change this signature to more clearly reflect/convey what's in the Entry by instead passing the key values? This reads much more clearly:
```
private void writeField(short thriftFieIdId, Object fieldValue, TProtocol oprot)
```